### ISSUE TITLE
n_pair_loss

### ISSUE THREAD

**Issue by tomp11** [leeesangwon/PyTorch-Image-Retrieval#11]
As the word "n_pair", batch-size is corretly 2n.
In your loss,  batch-size is n+1.
So, as written in this paper (http://www.nec-labs.com/uploads/images/Department-Images/MediaAnalytics/papers/nips16_npairmetriclearning.pdf),
you should chage your loss name into (N+1)-tuplet loss.

**Comment by leeesangwon** [comment]
In main.py, [BalancedBatchSampler](https://github.com/leeesangwon/PyTorch-Image-Retrieval/blob/039516da34dabee1d358ac15a2063cfb079d0bc1/main.py#L117) would make a mini-batch of size num_classes * num_samples.  
But, like you mentioned, n-pair loss only requires a mini-batch size of 2 * num_class.
Therefore when a user set more than 2 samples per class, [our n-pair loss implementation](https://github.com/leeesangwon/PyTorch-Image-Retrieval/blob/039516da34dabee1d358ac15a2063cfb079d0bc1/losses.py#L118) randomly chooses 2 samples and uses only them for backpropagation. So the actual mini-batch size would become 2 * num_classes.

Thanks.

**Comment by tomp11** [comment]
![npair](https://user-images.githubusercontent.com/36936262/60859421-8ccdf480-a24d-11e9-9161-0c42ce5ea86a.png)
I misunderstood that your loss created an n-pair each time as in the middle of the above figure.

However, I do not think that it is necessary to take each loss and then take an average.


```
def forward(self, f, f_p):
        n_pairs = len(f)
        term1 = 4 * self.sq_tan_alpha * torch.matmul(f + f_p, torch.transpose(f_p, 0, 1))
        term2 = 2 * (1 + self.sq_tan_alpha) * torch.sum(f * f_p, keepdim=True, dim=1)
        f_apn = term1 - term2
        mask = torch.ones_like(f_apn) - torch.eye(n_pairs).cuda()
        f_apn = f_apn * mask
        loss = torch.mean(torch.logsumexp(f_apn, dim=1))
```

I think that you can calculate the matrix at once like above, what do you think?

**Comment by leeesangwon** [comment]
Sorry for the late reply.
We'll be very grateful if you generate a PR of your idea to make us understand the suggestion more clearly.
Thank you again for your suggestion for improving our project.
